{% extends 'base.html' %}

{% block body %}
  <script>
  $( function() {
    $(document).tooltip();$('.accordion').asAccordion({namespace: 'accordion'});
    $('#FR_processed').prop('checked',false);
    tabs=$( "#tabs" );
    tabs.tabs();
    $('#HPara').hide();
    $('#KPara').hide();
    $('#tab3div').hide();$('#tab4div').hide();
    $('#my-select').multiSelect({ keepOrder: true });
    $('.downloadData1').hide();  
    $('#upload1').on('click',function(){
		form_data=new FormData();
		ins=document.getElementById('multiFiles').files.length;
		if(ins==0){
			$('#msg').html('<div class="alert-danger" role="alert">Select at least one file for Gene Expression</div>');
			return;
		}
		else{
		$('#msg').html('');
		}
		
		for(var i=0;i<ins;i++){
			form_data.append("files[]", document.getElementById('multiFiles').files[i]);
		}
		csrf_token=$('input[name="csrfmiddlewaretoken"]').val();
		form_data.append("csrfmiddlewaretoken",csrf_token);
		
		$.ajax({
			url: '/upload/geneExpression/',
			datType:'json',
			cache:false,
			contentType:false,
			processData:false,
			data:form_data,
			type:'post',
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#preview1').html(response);
				$('#preview2').html(' ');
				
			},
			error:function(response){
				alert('upload error!');
			}
		});
	});
	
   $('#upload2').on('click',function(){
		form_data=new FormData();
		ins=document.getElementById('singleFile').files.length;
		if(ins==0){
			$('#msg').html('<div class="alert-danger" role="alert">Select at least one file for Meta-data</div>');
			return;
		}
		else{
		$('#msg').html('');
		}
		
		form_data.append("meta", document.getElementById('singleFile').files[0]);
		csrf_token=$('input[name="csrfmiddlewaretoken"]').val();
		form_data.append("csrfmiddlewaretoken",csrf_token);
		
		$.ajax({
			url: '/upload/meta/',
			datType:'json',
			cache:false,
			contentType:false,
			processData:false,
			data:form_data,
			type:'post',
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#preview2').html(response);
				
			},
			error:function(response){
				alert('upload error!');
			}
		});
	});
	
   $('#slider1').slider({
		min:0,
		max:2,
		step:0.001,
		value:1,
		slide:function(event,ui){
			$('#sliderValue').val(ui.value);
		}
	});
	$('#sliderValue').val($('#slider1').slider('value'));
	$('#sliderValue').on('change',function(){
		$('#slider1').slider('value',$(this).val());
		$('#slider1').slider('refresh');
	});
	

	
   $('#next1').on('click',function(){	
		tabs.tabs('option','active',1);
	});
   $('#next2').on('click',function(){
   	$('.btn').attr('disabled',true);
		$.ajax({
			url: '/eda/?correct='+$('input[name="Comb"]:checked').val()+'&log2='+$('input[name="Log"]:checked').val()+'&fr='+$('input[name="FR"]:checked').val()+'&integrate='+$('#my-select').val(),
			datType:'html',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			
			success:function(response){
				$('#tab3div').show();
				tabs.tabs('option','active',2);
				dfs1=JSON.parse(response.dfs1);
				plotlyPlot('graphDiv1',dfs1);
				$('.btn').attr('disabled',false);
				dfs2=JSON.parse(response.dfs2);
				plotlyPlot('graphDiv2',dfs2);			
				$.ajax({
					url: '/dgea/',
					datType:'json',
					cache:false,
					contentType:false,
					processData:false,
					type:'get',
					async:true,
					complete:function(){
						$('#loader').hide();
					},
					success:function(response){
						$('#graphDiv11').html('<img class="img1" src="/static/temp/'+response[0]+'" width="800" height="auto"/>');
						$('#graphDiv21').html('<img class="img1" src="/static/temp/'+response[1]+'" width="800" height="auto"/>');
						alert('Calculation Succeed!');
						wheelzoom(document.querySelectorAll('img.img1'));
						//$('.accordion').asAccordion({namespace: 'accordion'});
					},
					error:function(response){
						alert(response.responseText);
						$('.btn').attr('disabled',false);
						$('#graphDiv11').html('');
						$('#graphDiv21').html('');
						return;
					}
				});
	
			},
			error:function(response){
				alert('server error!');
				$('.btn').attr('disabled',false);
				return;
			}
		});
		$('.btn').attr('disabled',false);
		
	});
   $('#next3').on('click',function(){
		tabs.tabs('option','active',3);
		$('#tab4div').hide();
	});
   $('#next4').on('click',function(){
		tabs.tabs('option','active',4);
	});

   $("input[type=radio][name=clustering]").on('change',function () {
            if($(this).val()=='LEIDEN')
            {
            	$('#HPara').hide();
            	$('#KPara').hide();
            	$('#LPara').show();
            }
            else if($(this).val()=='HDBSCAN'){
            	$('#LPara').hide();
            	$('#KPara').hide();
            	$('#HPara').show();
            }
            else{
            	$('#LPara').hide();
            	$('#HPara').hide();
            	$('#KPara').show();
            }
        }); 
  });
  
  
function unpack(rows,n){
	return rows.map(function(row)
	{return row[n];});}
	
function plotlyPlot(idP,dfsP){
	data=[];
	sequentialScale=d3.scaleSequential().domain([0,Object.keys(dfsP).length]).interpolator(d3.interpolateRainbow);
	i=0;
	for(const [key,value] of Object.entries(dfsP)){
		var trace={
		x:unpack(value.data, 0), y: unpack(value.data, 1), z: unpack(value.data, 2),
		mode: 'markers',
		marker: {
			size: 2,color:sequentialScale(i),
			symbol:'circle',
			opacity: 0.5},
		type: 'scatter3d',
		name: key,
		text: value.obs,
		};
		data.push(trace);
		i++;
	}
	var layout = {
		legend:{
	  	x:1,
	  	y:0.5,
		itemsizing: 'constant'
	  	}
	  };
	Plotly.newPlot(idP, data, layout);
}

function plotlyPlot1(idP,dfsP){
	data=[];
	sequentialScale=d3.scaleSequential().domain([0,Object.keys(dfsP).length]).interpolator(d3.interpolateRainbow);
	i=0;
	for(const [key,value] of Object.entries(dfsP)){
		var trace={x:value.x,y:value.y,name:value.name,type:value.type,marker:{color:sequentialScale(i)}};
		data.push(trace);
		i++;
	}
	layout={barmode:'stack'};
	Plotly.newPlot(idP,data,layout);
}

function clusterBtnClick(){
	if($('input[type=radio][name=clustering]:checked').val()=='LEIDEN')
        	url='/cluster/?cluster=LEIDEN&param='+$('#sliderValue').val();
        else if($('input[type=radio][name=clustering]:checked').val()=='HDBSCAN')
        	url='/cluster/?cluster=HDBSCAN&param='+$('#hdbP').val();
        else
        	url='/cluster/?cluster=Kmeans&param='+$('#kP').val();
        if($('#FR_processed').attr('checked'))
        	url=url+'&useFR=true';
        else
        	url=url+'&useFR=false';
        data=[];
        $('.btn').attr('disabled',true);
        $.ajax({
			url: url,
			datType:'html',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#tab4div').show();
				plotlyPlot('graphDiv1-cluster',response.traces);
				plotlyPlot1('bc1',response.bc1);
				plotlyPlot1('bc2',response.bc2);
				$('#graphDiv1-cluster-gene').html('<img class="img2" src="/static/temp/'+response.fileName+'" width="500" height="auto"/>');
				$('#graphDiv1-cluster-top-gene').html('<img class="img2" src="/static/temp/'+response.fileName1+'" width="1000" height="auto"/>');
				alert('Calculation Succeed!');
				wheelzoom(document.querySelectorAll('img.img2'));
				$('.downloadData1').show();
				$('#cluster_num').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#cluster_num1').html("<option value='' selected='selected'>----SELECT----</option>");
				for (let key in response.traces){
					$('#cluster_num').append(new Option(key,key));
					$('#cluster_num1').append(new Option(key,key));
				}
				$('.btn').attr('disabled',false);
			},
			error:function(response){
				alert(response.responseText);
				$('#graphDiv1-cluster-gene').html('');
				$('#graphDiv1-cluster-top-gene').html('');
				$('#cluster_num').html("<option value='' selected='selected'>----SELECT----</option>");
				$('#cluster_num1').html("<option value='' selected='selected'>----SELECT----</option>");
				$('.btn').attr('disabled',false);
				return;
			}
		});
	
}

function goEnrich(){
	$('.btn').attr('disabled',true);
	$.ajax({
			url:'/goenrich/?cluster_n='+$('#cluster_num').val(),
			datType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#goEnrichPic').html('<img class="imgGo" src="/static/temp/'+response.fileName+'" width="500" height="auto"/>');
				alert('Calculation Succeed!');
				wheelzoom(document.querySelectorAll('img.imgGo'));
				$('.btn').attr('disabled',false);
			},
			error:function(response){
				alert(response.responseText);
				$('#goEnrichPic').html('');
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function golasso(){
	$('.btn').attr('disabled',true);
	$.ajax({
			url:'/test/?cluster_n='+$('#cluster_num1').val(),
			datType:'json',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('#lassoPic').html('<img class="imgLasso" src="data:image/png;base64,'+response+'" width="1000" height="auto"/>');
				alert('Calculation Succeed!');
				wheelzoom(document.querySelectorAll('img.imgLasso'));
				$('.btn').attr('disabled',false);
			},
			error:function(response){
				alert(response.responseText);
				$('#lassoPic').html('');
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

function getTopGenes(t){
	url='';
	if(t==1){
		url='/dgea?clusters=fileName&topN='+$('#topForFiles').val();
	}
	else if(t==2){
		url='/dgea?clusters=label&topN='+$('#topForLabels').val();
	}
	else{
		url='/dgea?clusters='+$('input[type=radio][name=clustering]:checked').val()+'&topN='+$('#topForLabels').val();//LEIDEN,HDBSCAN,Kmeans
	}
	$.ajax({
			url:url,
			datType:'text',
			cache:false,
			contentType:false,
			processData:false,
			type:'get',
			async:true,
			beforeSend:function(){
			$('#loader').show();
			},
			complete:function(){
			$('#loader').hide();
			},
			success:function(response){
				$('.btn').attr('disabled',false);
				const blob = new Blob([response], { type: 'text/csv' });
   				 // Create a link element to trigger the download
				const link = document.createElement('a');
				link.href = window.URL.createObjectURL(blob);
				link.download = 'export.csv';

				// Append the link to the document and trigger a click event
				document.body.appendChild(link);
				link.click();

				// Remove the link from the document
				document.body.removeChild(link);
			},
			error:function(response){
				alert(response.responseText);
				$('.btn').attr('disabled',false);
				return;
			}
	});
}

  </script>
  
  <style>
  #loader{
  position:fixed;
  left:0px;
  top:0px;
  width:100%;
  height:100%;
  z-index:99999;
  background:url(/static/loading12.gif) 50% 50% no-repeat rgb(15 10 10 /59%);
  background-size:200px;
  }
  .rotated{
  transform:rotate(90deg);
  }

  </style>
  
<div id='loader' style='display:none'></div>
<div id="tabs" style='width:1800px;'>
  <ul>
    <li><a href="#tabs-1">Data Collection</a></li>
    <li><a href="#tabs-2">Data Processing</a></li>
    <li><a href="#tabs-3">EDA</a></li>
    <li><a href="#tabs-4">DEGA</a></li>
    <li><a href="#tabs-5">LASSO FeastureSelect</a></li>
    <li><a href="#tabs-6">Go Enrichment</a></li>
  </ul>
  <div id="tabs-1">
    <fieldset name="Multiple Files Upload">
    <div id="msg"></div>
    <p>Upload Gene Expression files&nbsp;<img id='tooltip1'class='question' title="In this section, users upload their files(multiple). File names will be used for later visualisation. ID_REF is a compulsory field for each file.<p><img src='/static/fileUpload.png' width='500' height='auto'/></p>"/>&nbsp;&nbsp;&nbsp;<button id="advancedSearch" class="btn btn-primary" onclick="window.open('/advancedSearch/');" title="This is used for processing any online data. User can search for any public dataset and integrate them into the research.">Advanced Search</button></p>
    <table>
    <tr>
	<td style="vertical-align:top;">
	{% csrf_token %}
		<input type="file" id="multiFiles" name="files[]" multiple="multiple" title="Choose from local"/>
		<button id="upload1" class="btn btn-primary">Upload</button>
	</td>

    </tr>
    <tr>
  	<td>
  	<div id='preview1' style="height:400px; width:1200px; overflow-x:scroll; overflow-y:scroll;"></div>	
  	</td>
    </tr>
    </table>
    
    <p>Upload Meta-data file&nbsp;<img class='question' title="In this section, users upload their meta-data file(single). ID_REF and LABEL are compulsory. ID_REF is matched with the previous ones in Expression files. LABEL will be used for later visualisation. User can upload clinic data as well.<p><img src='/static/metaUpload.png' width='auto' height='auto' /></p>"/></p>
    <table>
    <tr>
	<td style="vertical-align:top;">
	{% csrf_token %}
		<input type="file" id="singleFile" name="singleFile" title="Choose from local"/>
		<button id="upload2" class="btn btn-primary">Upload</button>
		<button id="next1" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>	
	</td>

    </tr>
    <tr>
  	<td>
  	<div id='preview2' style="height:400px; width:1200px; overflow-x:scroll; overflow-y:scroll;"></div>	
  	</td>
    </tr>
    </table>
   </fieldset>
  </div>
  
  <div id="tabs-2">
  <table>
  
  <tr>
  	<h5>Choose the built-in compiled cohort(s) you want to integrate with:</h5>
	<select id="my-select" multiple>
	<optgroup label='Shared-Cohorts'>
    	<option value="SERA" title="The Scotish Early Rheumatoid Arthritis;">SERA</option>
    	<option value="PEAC" title="The Pathobiology of Early Arthritis Cohort;">PEAC</option>
    	<option value="PSORT" title="Psoriasis Stratification to Optimise Relevant Therapy;">PSORT</option>
    	<option value="RAMAP" title="Rheumatoid Arthritis Map;">RAMAP</option>
		<option value="ORBIT" title="ORBIT was a study comparing Rituximab to Anti-TNF treatments. ">ORBIT</option>
    	</optgroup>
	</select>
  </tr>

  	
  <tr>
  	<td>
   	<div name="Options1" style="left:30px; width:250px;">
   		<br>
		<p>Combined Options</p>
		<input type="radio" name="Comb" value="Combat" checked="checked" /><label>&nbsp;Combat&nbsp;</label><img class='question' style="max-width:4%;" title="ComBat, originally implemented in the R library sva. This tool leverages a parametric and non-parametric empirical Bayes approach for correcting the batch effect in microarray datasets that works for small sample sizes or in the presence of outliers."/>
		<br>
  		<input type="radio" name="Comb" value="Harmony" /><label>&nbsp;Harmony&nbsp;</label><img class='question' style="max-width:4%;" title="<p>Harmony is a general-purpose package with an efficient algorithm for integrating multiple data sets. It is especially useful for large single-cell datasets such as single-cell RNA-seq.</p><p>Harmony is:</p><p><b>Fast:</b> Analyze thousands of cells on your laptop.</p><p><b>Sensitive:</b> Different cell types may be present or absent in each batch.</p><p><b>Accurate:</b> Integrate cells from multiple donors, tissues – even different technologies.</p>"/>
  		<br>
  		<br>
	</div>
	</td>
	
	<td>
	<div name="Options2" style="left:30px; width:250px;">
		<br>
		<p>Log2 Transformation</p>
		<input type="radio" name="Log" value="Yes" /><label>&nbsp;Yes</label>
		<br>
  		<input type="radio" name="Log" value="No" checked="checked"><label>&nbsp;No</label>
		<br>
		<br>
	</div>
	</td>

	<td>
	<div name="Options3" style="left:30px; width:250px;">
		<br>
		<p>Feature Reduction Method</p>
		<input type="radio" name="FR" value="TSNE" /><label>&nbsp;TSNE&nbsp;</label><img class='question' style="max-width:4%;" title="<p>t-SNE (t-distributed Stochastic Neighbor Embedding) is an unsupervised non-linear dimensionality reduction technique for data exploration and visualizing high-dimensional data. Non-linear dimensionality reduction means that the algorithm allows us to separate data that cannot be separated by a straight line. </p>"/>
		<br>
  		<input type="radio" name="FR" value="UMAP"  checked="checked" /><label>&nbsp;UMAP&nbsp;</label><img class='question' style="max-width:4%;" title="<p>Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique that can be used for visualisation similarly to t-SNE, but also for general non-linear dimension reduction.</p>"/>
		<br>
		<br>
	</div>
	</td>
  </tr>
  <tr>
  	<td colspan='2' style="text-align:right;">
	<button id="next2" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>
  	</td>
  </tr>
  </table>
  </div>
  
  
  <div id="tabs-3">
  	<p>Gene Data Explory Data Analyses</p>
  	<div id='tab3div'>	
          <ul data-direction="horizontal" data-multiple="true" data-initial-index="[0]" data-event="click" class="accordion accordion--box accordion--horizontal1">
            <li class="accordion__panel">
              <span class="accordion__heading">Data processed by Labels<i class="-icon -icon--left"></i></span>
              <div class="accordion__expander">
              		<div id="graphDiv2" style="width:850px;height:850px;overflow-x:scroll; overflow-y:scroll;"></div>
  			<div style="width:100%; border-top:1px solid #ccc;"></div>
  			<p>Dotplot of top 4 genes for each LABEL​</p>
  			<br><br><br><br><br><br><br><br><br>
  			<div id="graphDiv21"  class='rotated' style="width:800px;height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
			<br><br><br><br><br><br><br><br><br>
			<p>Customised Top Number of Genes for Download​:</p>
			<select id='topForLabels'>
					<option value='50' selected='selected'>50</option>
					<option value='100'>100</option>
					<option value='500'>500</option>
  			</select>
  			<button id='topForLabelsDownloadBtn' class='btn btn-primary' onclick='getTopGenes(2);'>Download Top Genes</button>
              </div>
            </li>
            <li class="accordion__panel">
              <span class="accordion__heading">Data processed by FileNames<i class="-icon -icon--left"></i></span>
              <div class="accordion__expander">
              		<div id="graphDiv1" style="width:850px;height:850px;overflow-x:scroll; overflow-y:scroll;"></div>
  			<div style="width:100%; border-top:1px solid #ccc;"></div>
  			<p>Dotplot of top 4 genes for each file​</p>
  			<br><br><br>
  			<div id="graphDiv11" class='rotated' style="width:600px;height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
			<br><br><br>
			<p>Customised Top Number for Download​:</p>
			<select id='topForFiles'>
					<option value='50' selected='selected'>50</option>
					<option value='100'>100</option>
					<option value='500'>500</option>
  			</select>
  			<button id='topForFilesDownloadBtn' class='btn btn-primary' onclick='getTopGenes(1);'>Download Top Genes</button>
            </div>
            </li>
          </ul>
  	</div>
  	<br>
	<div style="text-align:right;">
		<button id="download1" class="btn btn-primary" style="width:auto;" onclick="window.location='/static/temp/{{ user.get_username }}_corrected.csv';">Download Processed data</button>
		<br><br>
		<button id="next3" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>
			
	</div>

  </div>
 
 
   <div id="tabs-4">
  	<p>Choose Options for Clustering:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<input type='checkbox' id='FR_processed' title="Classify based on the TSNE/UMAP processed data">&nbsp;Classification based on Feature Reduction</input></p>
  	
	<table>
			<tr>
				<td style="height:50px; width:200px; border-right:1px solid #ccc;">
					<input type="radio" name='clustering' value='LEIDEN' checked /><label>&nbsp;LEIDEN&nbsp;</label><img class='question' style="max-width:6.5%;" title="<p>The Leiden algorithm is an algorithm for detecting communities in large networks. The algorithm separates nodes into disjoint communities so as to maximize a modularity score for each community. Modularity quantifies the quality of an assignment of nodes to communities, that is how densely connected nodes in a community are, compared to how connected they would be in a random network.<p></p>The Leiden algorithm is a hierarchical clustering algorithm, that recursively merges communities into single nodes by greedily optimizing the modularity and the process repeats in the condensed graph. It modifies the Louvain algorithm to address some of its shortcomings, namely the case where some of the communities found by Louvain are not well-connected. This is achieved by periodically randomly breaking down communities into smaller well-connected ones.</p>"/>
					<br>
					<input type="radio" name='clustering' value='HDBSCAN' /><label>&nbsp;HDBSCAN&nbsp;</label><img class='question' style="max-width:6.5%;" title="<p>HDBSCAN is a clustering algorithm developed by Campello, Moulavi, and Sander. It extends DBSCAN by converting it into a hierarchical clustering algorithm, and then using a technique to extract a flat clustering based in the stability of clusters. The goal of this notebook is to give you an overview of how the algorithm works and the motivations behind it. In contrast to the HDBSCAN paper I’m going to describe it without reference to DBSCAN. Instead I’m going to explain how I like to think about the algorithm, which aligns more closely with Robust Single Linkage with flat cluster extraction on top of it.</p>"/>
					<br>
					<input type="radio" name='clustering' value='KMeans' /></label>&nbsp;KMeans&nbsp;</label><img class='question' style="max-width:6.5%;" title="<p>k-means clustering is a method of vector quantization, originally from signal processing, that aims to partition n observations into k clusters in which each observation belongs to the cluster with the nearest mean (cluster centers or cluster centroid), serving as a prototype of the cluster. This results in a partitioning of the data space into Voronoi cells. k-means clustering minimizes within-cluster variances (squared Euclidean distances), but not regular Euclidean distances, which would be the more difficult Weber problem: the mean optimizes squared errors, whereas only the geometric median minimizes Euclidean distances. For instance, better Euclidean solutions can be found using k-medians and k-medoids.</p>"/>
				</td>	
				<td style="width:400px; height:50px; border-right:1px solid #ccc;">
					<div id='LPara'>
						<label for='slider1'>Input Param: Resolution  <img class='question' style="max-width:3%;" title="A parameter value controlling the coarseness of the clustering. Higher values lead to more clusters."/> = </label>
						<input type='text' id='sliderValue' size='6'></input>
						<br><br>
						<div id='slider1'></div>
					</div>
					<div id='HPara'>
						<label for='slider1'>Input Param: minSize  <img class='question' style="max-width:3%;" title="The minimum number of samples in a group for that group to be considered a cluster; groupings smaller than this size will be left as noise."/> = &nbsp;</label>
						<br>
						<input id='hdbP'  value='20' />
					</div>
					<div id='KPara'>
						<label for='slider1'>Input Param: K  <img class='question' style="max-width:3%;" title="The number of clusters to form as well as the number of centroids to generate."/> =  &nbsp; &nbsp;</label>
						<br>
						<input id='kP'  value='5' />
					</div>
				</td>
				<td style="height:50px;">
				<button id="clusterBtn" class="btn btn-primary" style="width:auto;" onclick="clusterBtnClick();">Cluster the Data</button>
				<button id='advanced' class="btn btn-primary" style='width:auto;' onclick="window.open('/cluster/advanced/');">Advanced</button>
				</td>
			</tr>
	</table>
	<div id='tab4div'>
	<ul data-direction="horizontal" data-multiple="true" data-initial-index="[0]" data-event="click" class="accordion accordion--box accordion--horizontal2">
            <li class="accordion__panel">
              <span class="accordion__heading">Data colored by clusters<i class="-icon -icon--left"></i></span>
              <div class="accordion__expander">
              		<div id='graphDiv1-cluster' style="width:850px; height:850px; overflow-x:scroll; overflow-y:scroll;"></div>
              </div>
            </li>
            <li class="accordion__panel">
              <span class="accordion__heading">Dotplot of top 4 genes for each cluster<i class="-icon -icon--left"></i></span>
              <div class="accordion__expander">
              		<br><br><br>
              		<div id="graphDiv1-cluster-gene" class='rotated' style="width:600px;height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
					<br><br><br>
					<p>Customised Top Number of Genes for Download​:</p>
						<select id='topForLabels'>
							<option value='50' selected='selected'>50</option>
							<option value='100'>100</option>
							<option value='500'>500</option>
						</select>
					<button id='topForLabelsDownloadBtn' class='btn btn-primary' onclick='getTopGenes(3);'>Download Top Genes</button>
              </div>
            </li>
        </ul>
	<table>
			<tr>
				<td><p>Contributions from each file</p></td>
				<td><p>Contributions from each LABEL</p></td>
			</tr>
			<tr>
				<td>
				<div id='bc1' style="width:600px; height:500px; overflow-x:scroll; overflow-y:scroll;"></div>
				</td>
				<td>	
				<div id="bc2" style="width:600px; height:500px;overflow-x:scroll; overflow-y:scroll;"></div>
				</td>
			</tr>
			<tr>
				<td colspan='2'>
				<p>Top 20 genes after running DE for clusters​</p>
				<div id='graphDiv1-cluster-top-gene' style="width:1000px; overflow-x:scroll; overflow-y:scroll;"></div>
				</td>
			</tr>
			<tr class='downloadData1'>
				<td style="text-align:right;" colspan='3'>
				<button id="download2" class="btn btn-primary" style="width:auto;" onclick="window.location='/static/temp/{{ user.get_username }}_adata.h5ad';">Download Processed data</button>
				<br><br>
					<button id="next4" class="btn btn-primary">&nbsp;&nbsp;Next&nbsp;&nbsp;</button>
				</td>
			</tr>
	</table>
	</div>	
	</table>
  </div>
  
  <div id='tabs-5'>
  	<p>Select one interesting cluster for Feature Selection by Lasso:&nbsp;<img class='question' style="max-width:0.8%;" title="<p>LASSO regression, also known as L1 regularization, is a popular technique used in statistical modeling and machine learning to estimate the relationships between variables and make predictions. LASSO stands for Least Absolute Shrinkage and Selection Operator.</p><p>The primary goal of LASSO regression is to find a balance between model simplicity and accuracy. It achieves this by adding a penalty term to the traditional linear regression model, which encourages sparse solutions where some coefficients are forced to be exactly zero. This feature makes LASSO particularly useful for feature selection, as it can automatically identify and discard irrelevant or redundant variables.</p>"/></p>
  	<table>
  		<tr>
			<td style="text-align:right;width:50%;">
			<select id='cluster_num1'>
			<option value='' selected='selected'>----SELECT----</option>
			</select>
			</td>
			<td style="text-align:left;width:50%;">
			<button id="cluster_btn" class="btn btn-primary" onclick="golasso();">Lasso</button>
			</td>
		</tr>
		<tr>
		<td colspan='2'>
		<div id='lassoPic' style="width:1200px; height:800px; overflow-x:scroll; overflow-y:scroll; text-align:center;"></div>
		</td>
		</tr>
  	</table>
  </div>
  
  <div id='tabs-6'>
  	<p>Select one interesting cluster for Go Enrichment Analysis:&nbsp;<img class='question' style="max-width:0.8%;" title="<p>GO (Gene Ontology) enrichment analysis is a computational method used in bioinformatics to identify overrepresented gene ontology terms associated with a set of genes. Gene Ontology provides a standardized vocabulary for annotating the molecular functions, biological processes, and cellular components of genes and gene products.</p>"/></p>
  	<table>
  		<tr>
			<td style="text-align:right;width:50%;">
			<select id='cluster_num'>
			<option value='' selected='selected'>----SELECT----</option>
			</select>
			</td>
			<td style="text-align:left;width:50%;">
			<button id="cluster_btn" class="btn btn-primary" onclick="goEnrich();">Go Enrichment</button>
			</td>
		</tr>
		<tr>
		<td colspan='2'>
		<div id='goEnrichPic' style="width:1400px; overflow-x:scroll; overflow-y:scroll; text-align:center;"></div>
		</td>
		</tr>
  	</table>
  </div>

{% endblock %}
